From 32646fca7b7425366fc1a5f0cf0bfb8cfaca5e82 Mon Sep 17 00:00:00 2001
From: Guido Cella <guido@guidocella.xyz>
Date: Mon, 14 Dec 2020 10:12:38 +0100
Subject: [PATCH] Allow centering the terminal

Adds a keybinding that toggles centering Alacritty horizontally when
it's the only window, while still tiling multiple windows.
This limits the width of long text making it easier to read, and avoids
covering the wallpaper more than necessary.
---
 config.def.h |  1 +
 dwl.c        | 17 +++++++++++++++++
 2 files changed, 18 insertions(+)

--- ./config.def.h.orig	2021-02-28 13:50:44.473112945 +0200
+++ ./config.def.h	2021-02-28 13:53:19.450738869 +0200
@@ -84,6 +84,7 @@
 	{ MODKEY,                    XKB_KEY_m,          setlayout,      {.v = &layouts[2]} },
 	{ MODKEY,                    XKB_KEY_space,      setlayout,      {0} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_space,      togglefloating, {0} },
+	{ MODKEY,                    XKB_KEY_z,          togglecenter,   {0} },
 	{ MODKEY, 					 XKB_KEY_e,    		togglefullscreen, {0} },
 	{ MODKEY,                    XKB_KEY_0,          view,           {.ui = ~0} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_parenright, tag,            {.ui = ~0} },
--- ./dwl.c.orig	2021-02-28 13:50:44.449779467 +0200
+++ ./dwl.c	2021-02-28 14:02:56.000978444 +0200
@@ -106,6 +106,7 @@
 	int bw;
 	unsigned int tags;
 	int isfloating;
+	bool centered;
 	uint32_t resize; /* configure serial of a pending resize */
 	int prevx;
 	int prevy;
@@ -286,6 +287,7 @@
 static void tag(const Arg *arg);
 static void tagmon(const Arg *arg);
 static void tile(Monitor *m);
+static void togglecenter(const Arg *arg);
 static void togglefloating(const Arg *arg);
 static void togglefullscreen(const Arg *arg);
 static void toggletag(const Arg *arg);
@@ -341,6 +343,8 @@
 static struct wl_list mons;
 static Monitor *selmon;
 
+static bool center;
+
 /* global event handlers */
 static struct wl_listener cursor_axis = {.notify = axisnotify};
 static struct wl_listener cursor_button = {.notify = buttonpress};
@@ -2272,6 +2276,10 @@
 		if (c->isfullscreen)
 			maximizeclient(c);
 		else if (i < m->nmaster) {
+            if (n == 1 && center && c->centered) {
+                resize(c, m->w.width / 4, m->w.y, m->w.width / 2, m->w.height - 2 * c->bw, 0);
+               return;
+            }
 			h = (m->w.height - my) / (MIN(n, m->nmaster) - i);
 			resize(c, m->w.x, m->w.y + my, mw, h, 0);
 			my += c->geom.height;
@@ -2285,6 +2293,13 @@
 }
 
 void
+togglecenter(const Arg *arg)
+{
+	center = !center;
+	tile(selmon);
+}
+
+void
 togglefloating(const Arg *arg)
 {
 	Client *sel = selclient();
